### PR TITLE
Fix data donation key submission (EXPOSUREAPP-5814)

### PR DESCRIPTION
**Pull request by chiljamgossow**
remove data only when new test registered to allow donation after removal (this may be a new requirement)

calculate data on registration as data may be overwritten by the time the contribution is being made
check for order of dates
check for missing data, occurs when data donation is enabled after test registration and before test result

Default values should be -1 if they cannot be calculated.

### CODE REVIEW

File: Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/common/PpaDataExtensionsTest.kt [hunk 1]
```diff
@@ -26,4 +30,69 @@ class PpaDataExtensionsTest : BaseTest() {
         PpaData.PPAFederalState.FEDERAL_STATE_SH.federalStateShortName shouldBe "SH"
         PpaData.PPAFederalState.FEDERAL_STATE_TH.federalStateShortName shouldBe "TH"
     }
+
+    @Test
+    fun `days since most recent date at risk level at test registration are calculated correctly`() {
+        val now = Instant.now()
+        val dayBeforeYesterdayAt2200 = now
```
Comment (LukasLechnerDev) [inline Corona-Warn-App/src/test/java/de/rki/coronawarnapp/datadonation/analytics/common/PpaDataExtensionsTest.kt:37]: Maybe use `Instant.parse("2021-03-15T22:00:00.000Z")` in these tests here to simplify creation of instants?

**Comment by LukasLechnerDev** [comment]
I enabled data donation in the onboarding, and then scanned a positive test and submitted my keys. 

This is the analytics data: 
```
client_metadata {
  android_api_level: 30
  app_config_e_tag: "\"5357f74807272abd40b4aed12cdc0e9b\""
  cwa_version {
    major: 1
    minor: 15
  }
  enf_version: 18210613000
}
exposure_risk_metadata_set {
  date_changed_compared_to_previous_submission: true
  most_recent_date_at_risk_level: -1
  risk_level_changed_compared_to_previous_submission: true
}
test_result_metadata_set {
  days_since_most_recent_date_at_risk_level_at_test_registration: -1
  hours_since_high_risk_warning_at_test_registration: -1
  risk_level_at_test_registration: RISK_LEVEL_LOW
  risk_level_at_test_registration_value: 2
  test_result: TEST_RESULT_POSITIVE
  test_result_value: 3
}
user_metadata {
}
```


and this is stored in analytics_key_sumission_localdata.xml:

```
<map>
    <long name="analytics_key_submission_testRegisteredAt" value="1615890580496" />
    <int name="analytics_key_submission_daysSinceMostRecentDateAtRiskLevelAtTestRegistration" value="-1" />
    <int name="analytics_key_submission_lastSubmissionFlowScreen" value="2" />
    <boolean name="analytics_key_submission_submitted" value="true" />
    <long name="analytics_key_submission_submittedAt" value="1615890590310" />
    <int name="analytics_key_submission_riskLevelAtTestRegistration" value="2" />
</map>
```

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionStorage.kt [hunk 2]
```diff
@@ -73,6 +73,11 @@ class AnalyticsKeySubmissionStorage @Inject constructor(
         defaultValue = -1
     )
 
+    val daysSinceMostRecentDateAtRiskLevelAtTestRegistration = prefs.createFlowPreference(
```
Comment (kolyaopahle) [inline Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionStorage.kt:76]: if you only ever access it via `.value` does this need to be a flow preference or would a get/set implementation work as well?

**Review by kolyaopahle — DISMISSED** [review]
one minor comment, nothing merge blocking, love the tests 🥇 👍

**Review by LukasLechnerDev — DISMISSED** [review]

File: Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionStorage.kt (hunk 2 above)
Comment (chiljamgossow) [reply Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/modules/keysubmission/AnalyticsKeySubmissionStorage.kt:76]: Yes, that's true. This was copy/paste convenience. Would still keep now as it does not hurt.

**Comment by chiljamgossow** [comment]
> I enabled data donation in the onboarding, and then scanned a positive test and submitted my keys.
> 
> This is the analytics data:
> 
> ```
> client_metadata {
>   android_api_level: 30
>   app_config_e_tag: "\"5357f74807272abd40b4aed12cdc0e9b\""
>   cwa_version {
>     major: 1
>     minor: 15
>   }
>   enf_version: 18210613000
> }
> exposure_risk_metadata_set {
>   date_changed_compared_to_previous_submission: true
>   most_recent_date_at_risk_level: -1
>   risk_level_changed_compared_to_previous_submission: true
> }
> test_result_metadata_set {
>   days_since_most_recent_date_at_risk_level_at_test_registration: -1
>   hours_since_high_risk_warning_at_test_registration: -1
>   risk_level_at_test_registration: RISK_LEVEL_LOW
>   risk_level_at_test_registration_value: 2
>   test_result: TEST_RESULT_POSITIVE
>   test_result_value: 3
> }
> user_metadata {
> }
> ```
> 
> and this is stored in analytics_key_sumission_localdata.xml:
> 
> ```
> <map>
>     <long name="analytics_key_submission_testRegisteredAt" value="1615890580496" />
>     <int name="analytics_key_submission_daysSinceMostRecentDateAtRiskLevelAtTestRegistration" value="-1" />
>     <int name="analytics_key_submission_lastSubmissionFlowScreen" value="2" />
>     <boolean name="analytics_key_submission_submitted" value="true" />
>     <long name="analytics_key_submission_submittedAt" value="1615890590310" />
>     <int name="analytics_key_submission_riskLevelAtTestRegistration" value="2" />
> </map>
> ```

Please test again. The key_submission_metadata_set must be present if data donation is turned on before receiving a positive test result and keys are submitted.

**Comment by LukasLechnerDev** [comment]
Tested it again: 

analytics_key_sumission_localdata.xml:
```
<map>
    <long name="analytics_key_submission_testRegisteredAt" value="1615969209626" />
    <int name="analytics_key_submission_daysSinceMostRecentDateAtRiskLevelAtTestRegistration" value="0" />
    <boolean name="analytics_key_submission_submitted" value="true" />
    <long name="analytics_key_submission_submittedAt" value="1615969225440" />
    <int name="analytics_key_submission_lastSubmissionFlowScreen" value="2" />
    <long name="analytics_key_submission_testResultReceivedAt" value="1615969209623" />
    <int name="analytics_key_submission_riskLevelAtTestRegistration" value="2" />
</map>
```

AnalyticsData:
```
# de.rki.coronawarnapp.server.protocols.internal.ppdd.PpaData$PPADataAndroid@5f6d86be
client_metadata {
  android_api_level: 30
  app_config_e_tag: "\"5357f74807272abd40b4aed12cdc0e9b\""
  cwa_version {
    major: 1
    minor: 15
  }
  enf_version: 18210613000
}
exposure_risk_metadata_set {
  date_changed_compared_to_previous_submission: true
  most_recent_date_at_risk_level: -1
  risk_level_changed_compared_to_previous_submission: true
}
key_submission_metadata_set {
  hours_since_high_risk_warning_at_test_registration: -1
  last_submission_flow_screen: SUBMISSION_FLOW_SCREEN_TEST_RESULT
  last_submission_flow_screen_value: 2
  submitted: true
}
test_result_metadata_set {
  hours_since_high_risk_warning_at_test_registration: -1
  risk_level_at_test_registration: RISK_LEVEL_LOW
  risk_level_at_test_registration_value: 2
  test_result: TEST_RESULT_POSITIVE
  test_result_value: 3
}
user_metadata {
}
```

`key_submission_metadata_set` is now included.

**Review by LukasLechnerDev — APPROVED** [review]

**Review by kolyaopahle — APPROVED** [review]

**Pull request merged**